<a href="https://colab.research.google.com/github/Oktavian19/2341720117_ML_2025/blob/main/JS09/TP_JS09_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [13]:
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Naive Bayes

In [14]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv('/content/drive/MyDrive/asset/ml/spam.csv', encoding='latin-1') # spesifiksi encoding diperlukan karena data tidak menggunakan UTF-8

df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [15]:
df = df.iloc[:, :2]
df.columns = ['label', 'text']
df = df.rename(columns={'v1': 'Labels', 'v2': 'SMS'})

df['label_num'] = df['label'].map({'ham': 0, 'spam': 1})

In [16]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(df['text'], df['label_num'], test_size=0.2, random_state=42)

print(f"Training shape: {X_train.shape}")
print(f"Testing shape: {X_test.shape}")

Training shape: (4457,)
Testing shape: (1115,)


## Fitur CountVectorizer

In [17]:
count_vect = CountVectorizer(stop_words='english')

# transf train & test
X_train_count = count_vect.fit_transform(X_train)
X_test_count = count_vect.transform(X_test)

# Train MultinomialNB
nb_count = MultinomialNB()
nb_count.fit(X_train_count, y_train)

# Predict
y_pred_count = nb_count.predict(X_test_count)

# Evaluate
acc_count = accuracy_score(y_test, y_pred_count)
print(f"Accuracy (CountVectorizer): {acc_count}")
print("Classification Report (CountVectorizer):")
print(classification_report(y_test, y_pred_count))

Accuracy (CountVectorizer): 0.9838565022421525
Classification Report (CountVectorizer):
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       965
           1       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



## Fitur TF-IDF

In [18]:
tf_idf = TfidfVectorizer(stop_words='english')

# transf train & test
X_train_tfidf = tf_idf.fit_transform(X_train)
X_test_tfidf = tf_idf.transform(X_test)

# Train MultinomialNB
nb_tfidf = MultinomialNB()
nb_tfidf.fit(X_train_tfidf, y_train)

# Predict
y_pred_tfidf = nb_tfidf.predict(X_test_tfidf)

# Evaluate
acc_tfidf = accuracy_score(y_test, y_pred_tfidf)
print(f"Accuracy (TF-IDF): {acc_tfidf}")
print("Classification Report (TF-IDF):")
print(classification_report(y_test, y_pred_tfidf))

Accuracy (TF-IDF): 0.9668161434977578
Classification Report (TF-IDF):
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       965
           1       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115



In [19]:
# Comparison
print("\n--- Comparison ---")
print(f"Accuracy CountVectorizer: {acc_count:.4f}")
print(f"Accuracy TF-IDF: {acc_tfidf:.4f}")


--- Comparison ---
Accuracy CountVectorizer: 0.9839
Accuracy TF-IDF: 0.9668


## Kesimpulan

Fitur CountVectorizer merupakan fitur yang lebih baik untuk mendeteksi spam pada dataset tersebut ketika dipasangkan dengna Multinominal Naive Bayes.

Meskipun TF-IDF menunjukkan nilai presisi di angka 100%, namun nilai recall jauh di bawahnya yaitu pada 75%. TF-IDF dinilai terlalu berhati-hati dalam mendeteksi pesan spam dan gagal mendeteksi seperempat pesan spam yang ada. Namun demikian, CountVectorizer dapat memberikan keselarasan pada skor presisi dan recall yaitu di angka 94% untuk nilai presisi dan 92% recall